### PR TITLE
[FLINK-24656][docs] Add documentation for window deduplication

### PR DESCRIPTION
**Pull request by beyond1920**
## What is the purpose of the change

The pr adds documentation for Window deduplication.

## Brief change log

Window-deduplication.md

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+Therefore, window Deduplication queries have better performance if users don't need results updated per record. Usually, Window Deduplication is used with [Windowing TVF]({{< ref "docs/dev/table/sql/queries/window-tvf" >}}) directly. Besides, Window Deduplication could be used with other operations based on [Windowing TVF]({{< ref "docs/dev/table/sql/queries/window-tvf" >}}), such as [Window Aggregation]({{< ref "docs/dev/table/sql/queries/window-agg" >}}), [Window TopN]({{< ref "docs/dev/table/sql/queries/window-topn">}}) and [Window Join]({{< ref "docs/dev/table/sql/queries/window-join">}}). 
+
+Window Deduplication can be defined in the same syntax as regular Deduplication, see [Deduplication documentation]({{< ref "docs/dev/table/sql/queries/deduplication" >}}) for more information.
+Besides that, Window Deduplication requires the `PARTITION BY` clause contains `window_start` and `window_end` columns of the relation applied [Windowing TVF]({{< ref "docs/dev/table/sql/queries/window-tvf" >}}) or [Window Aggregation]({{< ref "docs/dev/table/sql/queries/window-agg" >}}).
```
Comment (godfreyhe) [inline docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md:34]: only `Window Aggregation` is supported ?

File: docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+     ROW_NUMBER() OVER (PARTITION BY window_start, window_end [, col_key1...]
+       ORDER BY time_attr [asc|desc]) AS rownum
+   FROM table_name) -- relation applied windowing TVF
+WHERE rownum = 1 [AND conditions]
+```
+
+**Parameter Specification:**
+
+- `ROW_NUMBER()`: Assigns an unique, sequential number to each row, starting with one.
+- `PARTITION BY window_start, window_end [, col_key1...]`: Specifies the partition columns which contain `window_start` and `window_end` columns i.e. the window deduplicate key.
```
Comment (godfreyhe) [inline docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md:54]: `window_start` and `window_end` columns i.e. the window deduplicate key -> `window_start`, `window_end` and other partition keys ?

File: docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+The following shows the syntax of the Window Deduplication statement:
+
+```sql
+SELECT [column_list]
+FROM (
+   SELECT [column_list],
+     ROW_NUMBER() OVER (PARTITION BY window_start, window_end [, col_key1...]
+       ORDER BY time_attr [asc|desc]) AS rownum
+   FROM table_name) -- relation applied windowing TVF
+WHERE rownum = 1 [AND conditions]
```
Comment (godfreyhe) [inline docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md:48]: actually, `rownum <= 1` and `rownum < 2` are all accpted

File: docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+**Parameter Specification:**
+
+- `ROW_NUMBER()`: Assigns an unique, sequential number to each row, starting with one.
+- `PARTITION BY window_start, window_end [, col_key1...]`: Specifies the partition columns which contain `window_start` and `window_end` columns i.e. the window deduplicate key.
+- `ORDER BY time_attr [asc|desc]`: Specifies the ordering column, it must be a [time attribute]({{< ref "docs/dev/table/concepts/time_attributes" >}}). Currently Flink supports [processing time attribute]({{< ref "docs/dev/table/concepts/time_attributes" >}}#processing-time) and [event time attribute]({{< ref "docs/dev/table/concepts/time_attributes" >}}#event-time). Ordering by ASC means keeping the first row, ordering by DESC means keeping the last row.
+- `WHERE rownum = 1`: The `rownum = 1` is required for Flink to recognize this query is deduplication.
+
+{{< hint info >}}
+Note: the above pattern must be followed exactly, otherwise the optimizer won’t be able to translate the query.
```
Comment (godfreyhe) [inline docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md:59]: `the optimizer won’t be able to translate the query` -> `the optimizer won’t be able to translate the query to window deduplication`

File: docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+{{< /hint >}}
+
+## Example
+
+The following example shows how to keep last record for every tumbling 10 minutes window.
```
Comment (godfreyhe) [inline docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md:64]: 10 minutes tumbling window

File: docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+      ROW_NUMBER() OVER (PARTITION BY window_start, window_end ORDER BY bidtime DESC) as rownum
+    FROM TABLE(
+               TUMBLE(TABLE Bid, DESCRIPTOR(bidtime), INTERVAL '10' MINUTES))
+  ) WHERE rownum <= 1;
++------------------+-------+------+-------------+------------------+------------------+--------+
+|          bidtime | price | item | supplier_id |     window_start |       window_end | rownum |
++------------------+-------+------+-------------+------------------+------------------+--------+
+| 2020-04-15 08:09 |  5.00 |    D |   supplier4 | 2020-04-15 08:00 | 2020-04-15 08:10 |      1 |
+| 2020-04-15 08:17 |  6.00 |    F |   supplier5 | 2020-04-15 08:10 | 2020-04-15 08:20 |      1 |
++------------------+-------+------+-------------+------------------+------------------+--------+
```
Comment (godfreyhe) [inline docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md:79]: it's better we can provide the source table ddl and source table data, which could help users understand it more easily

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribution @beyond1920 , I left some comments

File: docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+# Window Deduplication
+{{< label Batch >}} {{< label Streaming >}}
```
[the remark this answers is not in the document]
Comment (beyond1920) [reply docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md:26]: Batch would be supported soon in [PR](https://github.com/apache/flink/pull/17666) and [PR](https://github.com/apache/flink/pull/17670). 
I mark it to be only streaming currently.

File: docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+Window Deduplication is a special [Deduplication]({{< ref "docs/dev/table/sql/queries/deduplication" >}}) which removes rows that duplicate over a set of columns, keeping only the first one or the last one for each window and other partitioned keys. 
+
+For streaming queries, unlike regular Deduplicate on continuous tables, window Deduplication does not emit intermediate results but only a final result at the end of the window. Moreover, window Deduplication purges all intermediate state when no longer needed.
+Therefore, window Deduplication queries have better performance if users don't need results updated per record. Usually, Window Deduplication is used with [Windowing TVF]({{< ref "docs/dev/table/sql/queries/window-tvf" >}}) directly. Besides, Window Deduplication could be used with other operations based on [Windowing TVF]({{< ref "docs/dev/table/sql/queries/window-tvf" >}}), such as [Window Aggregation]({{< ref "docs/dev/table/sql/queries/window-agg" >}}), [Window TopN]({{< ref "docs/dev/table/sql/queries/window-topn">}}) and [Window Join]({{< ref "docs/dev/table/sql/queries/window-join">}}).
```
[the remark this answers is not in the document]
Comment (beyond1920) [reply docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md:31]: not yet

File: docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md (hunk 1 above)
Comment (beyond1920) [reply docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md:34]: All window TVF operations are allowed

**Comment by beyond1920** [comment]
@godfreyhe Thanks a lot for review. I've updated the pull request, please have a look again.

File: docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+   FROM table_name) -- relation applied windowing TVF
+WHERE (rownum = 1 | rownum <=1 | rownum < 2) [AND conditions]
+```
+
+**Parameter Specification:**
+
+- `ROW_NUMBER()`: Assigns an unique, sequential number to each row, starting with one.
+- `PARTITION BY window_start, window_end [, col_key1...]`: Specifies the partition columns which contain `window_start`, `window_end` and other partition keys.
+- `ORDER BY time_attr [asc|desc]`: Specifies the ordering column, it must be a [time attribute]({{< ref "docs/dev/table/concepts/time_attributes" >}}). Currently Flink supports [processing time attribute]({{< ref "docs/dev/table/concepts/time_attributes" >}}#processing-time) and [event time attribute]({{< ref "docs/dev/table/concepts/time_attributes" >}}#event-time). Ordering by ASC means keeping the first row, ordering by DESC means keeping the last row.
+- `WHERE (rownum = 1 | rownum <=1 | rownum < 2)`: The `rownum = 1 | rownum <=1 | rownum < 2` is required for Flink to recognize this query is Window Deduplication.
```
Comment (godfreyhe) [inline docs/content.zh/docs/dev/table/sql/queries/window-deduplication.md:56]: `is required for Flink to recognize the first/last row is needed` ?

**Review by godfreyhe — COMMENTED** [review]
LGTM overall, and please correct the case of `window Deduplication`

**Review by godfreyhe — APPROVED** [review]
LGTM

**Pull request closed without merging**